File: algorithms/02_c/math/gaussian_elimination.c
/* Gaussian elimination with partial pivoting. */
#include <stdio.h>
#include <math.h>
int main(void) {
    double a[3][4] = {{2, 1, -1, 8}, {-3, -1, 2, -11}, {-2, 1, 2, -3}};
    for (int c = 0; c < 3; c++) {
        int p = c;
        for (int r = c + 1; r < 3; r++) if (fabs(a[r][c]) > fabs(a[p][c])) p = r;
        for (int k = 0; k < 4; k++) { double t = a[c][k]; a[c][k] = a[p][k]; a[p][k] = t; }
        for (int r = c + 1; r < 3; r++) {
            double f = a[r][c] / a[c][c];
            for (int k = c; k < 4; k++) a[r][k] -= f * a[c][k];
        }
    }
    double x[3];
    for (int r = 2; r >= 0; r--) {
        double s = a[r][3];
        for (int k = r + 1; k < 3; k++) s -= a[r][k] * x[k];
        x[r] = s / a[r][r];
    }
    if (fabs(x[0] - 2) > 1e-9 || fabs(x[1] - 3) > 1e-9 || fabs(x[2] + 1) > 1e-9) return 1;
    printf("x=%g %g %g\n", x[0], x[1], x[2]);
    return 0;
}

File: algorithms/11_golang/math/gaussian_elimination.go
// Gaussian elimination with partial pivoting.
package main

import (
	"fmt"
	"math"
)

func main() {
	a := [][]float64{{2, 1, -1, 8}, {-3, -1, 2, -11}, {-2, 1, 2, -3}}
	n := 3
	for c := 0; c < n; c++ {
		p := c
		for r := c + 1; r < n; r++ {
			if math.Abs(a[r][c]) > math.Abs(a[p][c]) {
				p = r
			}
		}
		a[c], a[p] = a[p], a[c]
		for r := c + 1; r < n; r++ {
			f := a[r][c] / a[c][c]
			for k := c; k <= n; k++ {
				a[r][k] -= f * a[c][k]
			}
		}
	}
	x := make([]float64, n)
	for r := n - 1; r >= 0; r-- {
		s := a[r][n]
		for k := r + 1; k < n; k++ {
			s -= a[r][k] * x[k]
		}
		x[r] = s / a[r][r]
	}
	if math.Abs(x[0]-2) > 1e-9 || math.Abs(x[1]-3) > 1e-9 || math.Abs(x[2]+1) > 1e-9 {
		panic("wrong solution")
	}
	fmt.Println("x=", x)
}
